### PR TITLE
test: add pact tests for courseware and course-home

### PR DESCRIPTION
**Pull request by Ali-D-Akbar**
[PROD-2469](https://openedx.atlassian.net/browse/PROD-2469)
Adding 5 more contract tests related to course-home and courseware and generating pact between frontend-app-learning and LMS. Interactions from both of the contract tests files will be merged into a single json pact file.

Steps to run these tests on your local machine:
1. npm install
2. npm test

### CODE REVIEW

File: src/course-home/data/pact-tests/lmsPact.test.jsx [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  log: path.resolve(process.cwd(), 'src/course-home/data/pact-tests/logs', 'pact.log'),
+  dir: path.resolve(process.cwd(), 'src/pacts'),
+  pactfileWriteMode: 'merge',
+  logLevel: 'DEBUG',
+  cors: true,
+});
+
+const courseId = 'course-v1:edX+DemoX+Demo_Course';
+const dateRegex = '^(?:[1-9]\\d{3}-(?:(?:0[1-9]|1[0-2])-(?:0[1-9]|1\\d|2[0-8])|(?:0[13-9]|1[0-2])-(?:29|30)|(?:0[13578]|1[02])-31)|(?:[1-9]\\d(?:0[48]|[2468][048]|[13579][26])|(?:[2468][048]|[13579][26])00)-02-29)T(?:[01]\\d|2[0-3]):[0-5]\\d:[0-5]\\d(?:Z|[+-][01]\\d:[0-5]\\d)$';
+const opaqueKeysRegex = '[\\w\\-~.:]';
```
Comment (DawoudSheraz) [inline src/course-home/data/pact-tests/lmsPact.test.jsx]: This is also present in the other pact test files. Let's create a test constants file and import them in the files

File: src/courseware/data/pact-tests/lmsPact.test.jsx [hunk 2]
```diff
@@ -434,4 +435,86 @@ describe('Courseware Service', () => {
...
+        withRequest: {
+          method: 'GET',
+          path: `/api/courseware/resume/${courseId}`,
+        },
+        willRespondWith: {
+          status: 200,
+          body: {
+            block_id: string('sample_block_id'),
+            section_id: string('sample_section_id'),
+            unit_id: string('sample_unit_id'),
```
Comment (DawoudSheraz) [inline src/courseware/data/pact-tests/lmsPact.test.jsx]: Please use appropriate ids instead of just sample ids. Use matchers to perform the validations

Comment (Ali-D-Akbar) [reply src/courseware/data/pact-tests/lmsPact.test.jsx]: These are just alphanumeric strings like `642fadf46d074aabb637f20af320fb31`.
Should I add some data like this?

File: src/course-home/data/pact-tests/lmsPact.test.jsx [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        willRespondWith: {
+          status: 200,
+          body: {
+            can_show_upgrade_sock: boolean(false),
+            verified_mode: like({
+              access_expiration_date: null,
+              currency: 'USD',
+              currency_symbol: '$',
+              price: 149,
+              sku: '8CF08E5',
```
Comment (DawoudSheraz) [inline src/course-home/data/pact-tests/lmsPact.test.jsx:58]: use appropriate matchers

File: src/course-home/data/pact-tests/lmsPact.test.jsx [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+              upgrade_url: `${getConfig().ECOMMERCE_BASE_URL}/basket/add/?sku=8CF08E5`,
+            }),
+            can_load_courseware: boolean(true),
+            celebrations: {
```
Comment (DawoudSheraz) [inline src/course-home/data/pact-tests/lmsPact.test.jsx]: use like with the dicts everywhere to enforce the validation checks during provider verification

File: src/course-home/data/pact-tests/lmsPact.test.jsx [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+              verified_upgrade_link: `${getConfig().ECOMMERCE_BASE_URL}/basket/add/?sku=8CF08E5`,
+            }),
+            course_date_blocks: eachLike({
+              assignment_type: null,
+              complete: null,
+              date: term({
+                generate: '2013-02-05T05:00:00Z',
+                matcher: dateRegex,
+              }),
+              date_type: 'verified-upgrade-deadline',
```
Comment (DawoudSheraz) [inline src/course-home/data/pact-tests/lmsPact.test.jsx]: what other date types can be there(used by learning MFE)?

File: src/courseware/data/pact-tests/lmsPact.test.jsx [hunk 6]
```diff
@@ -181,14 +185,14 @@ describe('Courseware Service', () => {
             original_user_is_staff: boolean(true),
             can_view_legacy_courseware: boolean(true),
             is_staff: boolean(true),
-            course_access: like({
+            course_access: {
```
Comment (DawoudSheraz) [inline src/courseware/data/pact-tests/lmsPact.test.jsx]: why was like removed from here? Ideally, we would want to enforce the possible dict structure during provider verification

File: src/courseware/data/pact-tests/lmsPact.test.jsx (hunk 2 above)
Comment (DawoudSheraz) [reply src/courseware/data/pact-tests/lmsPact.test.jsx]: yes

File: src/courseware/data/pact-tests/lmsPact.test.jsx [hunk 7]
```diff
@@ -434,4 +431,51 @@ describe('Courseware Service', () => {
...
+      const response = await getResumeBlock(courseId);
+      expect(response).toBeTruthy();
+      expect(response).toEqual(camelCaseResponse);
+    });
+  });
+
+  describe('When a request to send activation email is made', () => {
+    it('returns status code 200', async () => {
+      await provider.addInteraction({
+        state: 'A logged-in user may or may not be active',
```
Comment (DawoudSheraz) [inline src/courseware/data/pact-tests/lmsPact.test.jsx:470]: First test for non-active users only. We can have a separate interaction for active users if needed.

**Review by DawoudSheraz — COMMENTED** [review]
left some comments to address.

File: src/course-home/data/pact-tests/lmsPact.test.jsx (hunk 5 above)
Comment (Ali-D-Akbar) [reply src/course-home/data/pact-tests/lmsPact.test.jsx]: There are total 9 date types. (from lms/djangoapps/courseware/date_summary.py)
'event', 'todays-date', 'course-start-date', 'course-end-date', 'assignment-due-date', 'course-expired-date', 'certificate-available-date', 'verified-upgrade-deadline', 'verification-deadline-date'

Comment (DawoudSheraz) [reply src/course-home/data/pact-tests/lmsPact.test.jsx]: then use regex/matchers for this

**Review by DawoudSheraz — APPROVED** [review]

**Review by azanbinzahid — APPROVED** [review]

**Comment by robrap** [comment]
@openedx/incident-management: Would Pact still work as a dev-only dependency? There is an issue with Tutor, Pact, and ARM64 systems.

**Comment by kdmccormick** [comment]
If Pact can in fact be a dev-only dependency, I have a draft PR that would make that change: https://github.com/openedx/frontend-app-learning/pull/823

**Comment by DawoudSheraz** [comment]
@robrap @kdmccormick Hello. Yes, this should be a dev-only dependency as it is used in the tests only. Inclusion in main dependencies was incorrectly done from our end.

**Comment by kdmccormick** [comment]
Thanks @DawoudSheraz , could you review https://github.com/openedx/frontend-app-learning/pull/823?

**Pull request merged**